File: mobile/lib/services/price_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

/// Price data from Indodax including 24h changes
class IndodaxTicker {
  final double last;      // Last traded price
  final double buy;       // Best buy price
  final double sell;      // Best sell price
  final double high;      // 24h high
  final double low;       // 24h low
  final double volume;    // 24h volume in IDR
  final String name;      // Token name

  IndodaxTicker({
    required this.last,
    required this.buy,
    required this.sell,
    required this.high,
    required this.low,
    required this.volume,
    required this.name,
  });

  factory IndodaxTicker.fromJson(Map<String, dynamic> json) {
    return IndodaxTicker(
      last: double.tryParse(json['last']?.toString() ?? '0') ?? 0,
      buy: double.tryParse(json['buy']?.toString() ?? '0') ?? 0,
      sell: double.tryParse(json['sell']?.toString() ?? '0') ?? 0,
      high: double.tryParse(json['high']?.toString() ?? '0') ?? 0,
      low: double.tryParse(json['low']?.toString() ?? '0') ?? 0,
      volume: double.tryParse(json['vol_idr']?.toString() ?? '0') ?? 0,
      name: json['name']?.toString() ?? '',
    );
  }
}

class PriceService {
  // Backend proxy for Indodax prices (phone can't reach Indodax directly)
  static const String _backendPriceApi = 'http://203.194.112.143:3000/prices';
  
  // Fallback: Direct Indodax API (if backend unavailable)
  static const String _indodaxApi = 'https://indodax.com/api';
  
  // Token pair mapping for Indodax (symbol -> ticker_id format)
  // Based on /api/pairs response, ticker_id uses underscore format like "btc_idr"
  static const Map<String, String> _indodaxPairs = {
    'POL': 'pol_idr',       // Polygon (POL)
    'MATIC': 'matic_idr',   // Polygon (legacy MATIC)
    'ETH': 'eth_idr',       // Ethereum
    'BTC': 'btc_idr',       // Bitcoin
    'USDT': 'usdt_idr',     // Tether
    'USDC': 'usdc_idr',     // USD Coin
    'LSK': 'lsk_idr',       // Lisk
    'SOL': 'sol_idr',       // Solana
    'BNB': 'bnb_idr',       // Binance Coin
    'XRP': 'xrp_idr',       // Ripple
    'DOGE': 'doge_idr',     // Dogecoin
    'ADA': 'ada_idr',       // Cardano
    'DOT': 'dot_idr',       // Polkadot
    'LINK': 'link_idr',     // Chainlink
    'UNI': 'uni_idr',       // Uniswap
    'AVAX': 'avax_idr',     // Avalanche
    'SHIB': 'shib_idr',     // Shiba Inu
  };
  
  // Fallback prices in IDR (updated regularly)
  static const Map<String, double> _fallbackPricesIDR = {
    'POL': 7500,
    'MATIC': 7500,
    'ETH': 58500000,
    'BTC': 1586500000,
    'LSK': 3200,
    'USDT': 16700,
    'USDC': 16700,
    'SOL': 3000000,
    'BNB': 10000000,
    'XRP': 31950,
    'DOGE': 2245,
    'ADA': 6200,
    'DOT': 33500,
    'LINK': 218000,
    'UNI': 99350,
    'AVAX': 223000,
    'SHIB': 0.37,     // Fixed SHIB price in IDR (was showing 0)
  };

  // Cache
  static Map<String, double> _priceCacheIDR = {};
  static Map<String, IndodaxTicker> _tickerCache = {};
  static Map<String, double> _prices24hAgo = {};
  static DateTime? _lastFetch;
  static DateTime? _lastBackendFailure;
  static const Duration _cacheDuration = Duration(seconds: 30); // Cache 30 detik untuk mengurangi spam
  static const Duration _backendCooldown = Duration(minutes: 1); // Cooldown setelah backend gagal
  
  // USD to IDR rate
  static const double _usdToIdr = 16700;

  final http.Client _httpClient;

  PriceService() : _httpClient = http.Client();

  /// Fetch all prices from Backend (which proxies to Indodax)
  /// Backend can reach Indodax, phone may not be able to
  /// Updates every 5 seconds for accurate LSK/IDR conversion
  Future<Map<String, double>> fetchPricesIDR(List<String> symbols) async {
    // Check cache (5 seconds)
    if (_lastFetch != null && 
        DateTime.now().difference(_lastFetch!) < _cacheDuration &&
        _priceCacheIDR.isNotEmpty) {
      final cached = <String, double>{};
      bool allFound = true;
      for (final symbol in symbols) {
        if (_priceCacheIDR.containsKey(symbol.toUpperCase())) {
          cached[symbol.toUpperCase()] = _priceCacheIDR[symbol.toUpperCase()]!;
        } else {
          allFound = false;
        }
      }
      if (allFound) return cached;
    }

    final prices = <String, double>{};

    // Skip backend if in cooldown after failure
    final shouldSkipBackend = _lastBackendFailure != null && 
        DateTime.now().difference(_lastBackendFailure!) < _backendCooldown;
    
    // Try backend first (backend can reach Indodax)
    if (!shouldSkipBackend) {
      try {
        final response = await _httpClient.get(
          Uri.parse(_backendPriceApi),
        ).timeout(const Duration(seconds: 5));

        if (response.statusCode == 200) {
          final data = json.decode(response.body);
          if (data['success'] == true && data['prices'] != null) {
            final backendPrices = data['prices'] as Map<String, dynamic>;
            
            for (final entry in backendPrices.entries) {
              final symbol = entry.key.toUpperCase();
              final price = (entry.value is num) ? entry.value.toDouble() : double.tryParse(entry.value.toString()) ?? 0;
              if (price > 0) {
                prices[symbol] = price;
              }
            }
            
            // Update cache
            _priceCacheIDR = {..._priceCacheIDR, ...prices};
            _lastFetch = DateTime.now();
            _lastBackendFailure = null; // Reset failure state on success
          }
        }
      } catch (e) {
        // Set cooldown to avoid spamming failed backend
        _lastBackendFailure = DateTime.now();
        print('[PriceService] Backend proxy failed, using fallback for 1 minute');
      }
    }
    
    // Fallback: Try direct Indodax if no prices yet
    if (prices.isEmpty) {
      try {
        final response = await _httpClient.get(
          Uri.parse('$_indodaxApi/summaries'),
        ).timeout(const Duration(seconds: 10));

        if (response.statusCode == 200) {
          final data = json.decode(response.body);
          final tickers = data['tickers'] as Map<String, dynamic>?;
          
          if (tickers != null) {
            for (final entry in _indodaxPairs.entries) {
              final symbol = entry.key;
              final pair = entry.value;
              final ticker = tickers[pair];
              if (ticker != null) {
                final tickerData = IndodaxTicker.fromJson(ticker);
                _tickerCache[symbol] = tickerData;
                prices[symbol] = tickerData.last;
              }
            }
          }
          
          _priceCacheIDR = {..._priceCacheIDR, ...prices};
          _lastFetch = DateTime.now();
        }
      } catch (e2) {
        // Silent fail - will use fallback prices
      }
    }
    
    // Use fallback for missing symbols
    for (final symbol in symbols) {
      final upperSymbol = symbol.toUpperCase();
      if (!prices.containsKey(upperSymbol) && _fallbackPricesIDR.containsKey(upperSymbol)) {
        prices[upperSymbol] = _fallbackPricesIDR[upperSymbol]!;
        // Disabled verbose fallback logging
        // if (Env.enableDebugLogs) print('[PriceService] Fallback $upperSymbol: Rp ${_fallbackPricesIDR[upperSymbol]}');
      }
    }

    return prices;
  }

  /// Get 24h price change percentage for a symbol
  double get24hChangePercent(String symbol) {
    final current = _priceCacheIDR[symbol.toUpperCase()];
    final ago = _prices24hAgo[symbol.toUpperCase()];
    if (current == null || ago == null || ago == 0) return 0;
    return ((current - ago) / ago) * 100;
  }

  /// Get full ticker data for a symbol
  IndodaxTicker? getTicker(String symbol) {
    return _tickerCache[symbol.toUpperCase()];
  }

  /// Fetch prices in USD (converted from IDR)
  Future<Map<String, double>> fetchPricesUSD(List<String> symbols) async {
    final idrPrices = await fetchPricesIDR(symbols);
    
    final usdPrices = <String, double>{};
    for (final entry in idrPrices.entries) {
      usdPrices[entry.key] = entry.value / _usdToIdr;
    }
    
    return usdPrices;
  }

  double get usdToIdr => _usdToIdr;

  /// Clear all cached data (call on logout)
  static void clearCache() {
    _priceCacheIDR.clear();
    _tickerCache.clear();
    _prices24hAgo.clear();
    _lastFetch = null;
  }

  Future<TotalBalanceResult> calculateTotalIDR(List<TokenBalanceInput> balances) async {
    final symbols = balances.map((b) => b.symbol).toList();
    final pricesUSD = await fetchPricesUSD(symbols);
    
    double totalUSD = 0;
    final tokens = <TokenWithPrice>[];

    for (final balance in balances) {
      final priceUSD = pricesUSD[balance.symbol.toUpperCase()] ?? 0;
      final priceIDR = priceUSD * _usdToIdr;
      final valueUSD = balance.balance * priceUSD;
      final valueIDR = balance.balance * priceIDR;
      
      totalUSD += valueUSD;
      
      tokens.add(TokenWithPrice(
        symbol: balance.symbol,
        balance: balance.balance,
        priceUSD: priceUSD,
        priceIDR: priceIDR,
        valueUSD: valueUSD,
        valueIDR: valueIDR,
      ));
    }

    return TotalBalanceResult(
      tokens: tokens,
      totalUSD: totalUSD,
      totalIDR: totalUSD * _usdToIdr,
    );
  }

  void dispose() {
    _httpClient.close();
  }
}

class TokenBalanceInput {
  final String symbol;
  final double balance;

  TokenBalanceInput({required this.symbol, required this.balance});
}

class TokenWithPrice {
  final String symbol;
  final double balance;
  final double priceUSD;
  final double priceIDR;
  final double valueUSD;
  final double valueIDR;

  TokenWithPrice({
    required this.symbol,
    required this.balance,
    required this.priceUSD,
    required this.priceIDR,
    required this.valueUSD,
    required this.valueIDR,
  });
}

class TotalBalanceResult {
  final List<TokenWithPrice> tokens;
  final double totalUSD;
  final double totalIDR;

  TotalBalanceResult({
    required this.tokens,
    required this.totalUSD,
    required this.totalIDR,
  });

  String get totalIDRFormatted {
    if (totalIDR >= 1000000) {
      return 'Rp ${(totalIDR / 1000000).toStringAsFixed(2)} Jt';
    } else if (totalIDR >= 1000) {
      return 'Rp ${(totalIDR / 1000).toStringAsFixed(0)} Rb';
    }
    return 'Rp ${totalIDR.toStringAsFixed(0)}';
  }

  String get totalIDRFullFormatted {
    final formatted = totalIDR.toStringAsFixed(0);
    // Add thousand separators
    final chars = formatted.split('').reversed.toList();
    final result = <String>[];
    for (int i = 0; i < chars.length; i++) {
      if (i > 0 && i % 3 == 0) result.add('.');
      result.add(chars[i]);
    }
    return 'Rp ${result.reversed.join('')}';
  }
}

File: backend/src/services/priceService.ts
import axios from "axios";

interface PriceCache {
  prices: Record<string, number>;
  pricesIDR: Record<string, number>;
  lastUpdated: number;
  indodaxLastUpdated: number;
}

const INDODAX_API = "https://indodax.com/api";
const COINCAP_API = "https://api.coincap.io/v2";
const CACHE_TTL = 1000; // 1 second - real-time price updates

// Indodax pair mapping (symbol -> ticker_id)
const INDODAX_PAIRS: Record<string, string> = {
  LSK: 'lsk_idr',
  ETH: 'eth_idr',
  BTC: 'btc_idr',
  USDT: 'usdt_idr',
  USDC: 'usdc_idr',
  POL: 'pol_idr',
  MATIC: 'matic_idr',
  SOL: 'sol_idr',
  BNB: 'bnb_idr',
  XRP: 'xrp_idr',
  DOGE: 'doge_idr',
  ADA: 'ada_idr',
  DOT: 'dot_idr',
  LINK: 'link_idr',
  UNI: 'uni_idr',
  AVAX: 'avax_idr',
  SHIB: 'shib_idr',
};

// Token ID mapping for CoinCap (fallback)
const TOKEN_IDS: Record<string, string> = {
  POL: "matic-network",
  MATIC: "matic-network", 
  ETH: "ethereum",
  BTC: "bitcoin",
  USDT: "tether",
  USDC: "usd-coin",
  LSK: "lisk",
};

let priceCache: PriceCache = {
  prices: {},
  pricesIDR: {},
  lastUpdated: 0,
  indodaxLastUpdated: 0,
};

// USD to IDR rate
let usdToIdr = 16700;

export async function updateUsdToIdr(): Promise<void> {
  try {
    // Use a free forex API or hardcode approximate rate
    const response = await axios.get(
      "https://api.exchangerate-api.com/v4/latest/USD",
      { timeout: 5000 }
    );
    if (response.data?.rates?.IDR) {
      usdToIdr = response.data.rates.IDR;
    }
  } catch (e) {
    // Disabled verbose logging
    // console.log("[PriceService] Using default USD/IDR rate:", usdToIdr);
  }
}

// Fallback prices in IDR (updated for current market)
const FALLBACK_PRICES_IDR: Record<string, number> = {
  LSK: 3400,
  ETH: 58500000,
  BTC: 1500000000,
  POL: 7500,
  MATIC: 7500,
  USDT: 16700,
  USDC: 16700,
  SOL: 2150000,
  BNB: 14700000,
  XRP: 32000,
  DOGE: 2250,
  ADA: 6200,
  DOT: 33500,
  LINK: 218000,
  UNI: 99350,
  AVAX: 223000,
  SHIB: 0.37,
};

// Fallback prices in USD
const FALLBACK_PRICES: Record<string, number> = {
  LSK: 0.20,
  ETH: 3500,
  BTC: 95000,
  POL: 0.45,
  MATIC: 0.45,
  USDT: 1.0,
  USDC: 1.0,
};

/**
 * Fetch prices from Indodax API (IDR native - most accurate for Indonesia)
 * Updates every 5 seconds for real-time conversion
 */
export async function fetchPricesFromIndodax(): Promise<Record<string, number>> {
  const now = Date.now();
  
  // Return cached if still valid
  if (now - priceCache.indodaxLastUpdated < CACHE_TTL && Object.keys(priceCache.pricesIDR).length > 0) {
    return priceCache.pricesIDR;
  }

  const pricesIDR: Record<string, number> = {};
  
  try {
    // Disabled verbose logging for troubleshooting
    // console.log("[PriceService] Fetching prices from Indodax API...");
    const response = await axios.get(`${INDODAX_API}/summaries`, {
      timeout: 10000,
    });

    if (response.data?.tickers) {
      const tickers = response.data.tickers;
      
      for (const [symbol, pair] of Object.entries(INDODAX_PAIRS)) {
        const ticker = tickers[pair];
        if (ticker?.last) {
          const price = parseFloat(ticker.last);
          if (price > 0) {
            pricesIDR[symbol] = price;
            // Disabled verbose per-coin logging
            // console.log(`[PriceService] ${symbol}: Rp ${price.toLocaleString()} (Indodax)`);
          }
        }
      }
    }

    // Update cache
    priceCache.pricesIDR = { ...priceCache.pricesIDR, ...pricesIDR };
    priceCache.indodaxLastUpdated = now;
    
    // Disabled verbose logging
    // console.log(`[PriceService] Updated ${Object.keys(pricesIDR).length} prices from Indodax`);
  } catch (e: any) {
    console.warn("[PriceService] Indodax API failed:", e.message);
    // Return cached or fallback
    if (Object.keys(priceCache.pricesIDR).length > 0) {
      return priceCache.pricesIDR;
    }
  }

  // Use fallback for missing symbols
  for (const [symbol, fallbackPrice] of Object.entries(FALLBACK_PRICES_IDR)) {
    if (!pricesIDR[symbol]) {
      pricesIDR[symbol] = fallbackPrice;
    }
  }

  return pricesIDR;
}

export async function fetchPrices(symbols: string[]): Promise<Record<string, number>> {
  const now = Date.now();
  
  // Return cached if still valid
  if (now - priceCache.lastUpdated < CACHE_TTL) {
    const cached: Record<string, number> = {};
    for (const symbol of symbols) {
      if (priceCache.prices[symbol.toUpperCase()]) {
        cached[symbol.toUpperCase()] = priceCache.prices[symbol.toUpperCase()];
      }
    }
    if (Object.keys(cached).length === symbols.length) {
      return cached;
    }
  }

  const prices: Record<string, number> = {};
  
  // Try to fetch from CoinCap API
  try {
    const response = await axios.get(`${COINCAP_API}/assets`, {
      params: { limit: 100 },
      timeout: 10000,
    });

    if (response.data?.data) {
      for (const asset of response.data.data) {
        for (const [symbol, coinCapId] of Object.entries(TOKEN_IDS)) {
          if (asset.id === coinCapId && asset.priceUsd) {
            prices[symbol] = parseFloat(asset.priceUsd);
          }
        }
      }
    }
  } catch (e: any) {
    console.warn("[PriceService] CoinCap API failed:", e.message);
  }

  // Use fallback for missing symbols
  for (const symbol of symbols) {
    const upper = symbol.toUpperCase();
    if (!prices[upper] && FALLBACK_PRICES[upper]) {
      prices[upper] = FALLBACK_PRICES[upper];
    }
  }

  // Update cache
  priceCache.prices = { ...priceCache.prices, ...prices };
  priceCache.lastUpdated = now;

  return prices;
}

export async function fetchPricesIDR(symbols: string[]): Promise<Record<string, number>> {
  const usdPrices = await fetchPrices(symbols);
  const idrPrices: Record<string, number> = {};

  for (const [symbol, usdPrice] of Object.entries(usdPrices)) {
    idrPrices[symbol] = usdPrice * usdToIdr;
  }

  priceCache.pricesIDR = idrPrices;
  return idrPrices;
}

export function getUsdToIdr(): number {
  return usdToIdr;
}

export async function getTokenPriceUSD(symbol: string): Promise<number> {
  const prices = await fetchPrices([symbol]);
  return prices[symbol.toUpperCase()] || 0;
}

export async function getTokenPriceIDR(symbol: string): Promise<number> {
  const prices = await fetchPricesIDR([symbol]);
  return prices[symbol.toUpperCase()] || 0;
}

export interface TokenWithPrice {
  symbol: string;
  balance: number;
  priceUSD: number;
  priceIDR: number;
  valueUSD: number;
  valueIDR: number;
}

export async function calculateTotalBalanceIDR(
  balances: Array<{ symbol: string; balance: number }>
): Promise<{
  tokens: TokenWithPrice[];
  totalUSD: number;
  totalIDR: number;
}> {
  const symbols = balances.map((b) => b.symbol);
  const pricesUSD = await fetchPrices(symbols);
  
  let totalUSD = 0;
  const tokens: TokenWithPrice[] = [];

  for (const { symbol, balance } of balances) {
    const priceUSD = pricesUSD[symbol.toUpperCase()] || 0;
    const priceIDR = priceUSD * usdToIdr;
    const valueUSD = balance * priceUSD;
    const valueIDR = balance * priceIDR;
    
    totalUSD += valueUSD;
    
    tokens.push({
      symbol,
      balance,
      priceUSD,
      priceIDR,
      valueUSD,
      valueIDR,
    });
  }

  return {
    tokens,
    totalUSD,
    totalIDR: totalUSD * usdToIdr,
  };
}

// Initialize USD/IDR rate on startup
updateUsdToIdr();
